### ISSUE TITLE
Own Home with Search Guard

### ISSUE THREAD

**Issue by Poohowy** [wtakase/kibana-own-home#16]
Hi. Does anyone know if Own Home plugin work correctly with Search Guard 5? Is nginx required? Do I have to change something in Kibana/Elastic/SG/Own Home config files? Can anyone could give me example of Own Home implementation? I will be really grateful for the response.

**Comment by wtakase** [comment]
I home [this](https://github.com/wtakase/kibana-own-home/issues/12#issuecomment-288314242) may help you.

**Comment by Poohowy** [comment]
This was really helpful, but I have some issues with Own Home. I think I made a mistake in the configuration. Can you check my conf files?

I check Search Guard and it works fine.
I have no LDAP, I use SG internal user database.

ES version: 5.1.2
Kibana version: 5.1.2
SG version: 5

sg_action_groups.yml:
```
KIBANA_READ:
  - READ
  - "indices:admin/mappings/fields/get*"
  - "indices:admin/validate/query*"
  - "indices:admin/get*"
```

sg_roles.yml:
```
sg_kibana_server:
  cluster:
      - CLUSTER_MONITOR
      - CLUSTER_COMPOSITE_OPS
  indices:
    '?kibana*':
      '*':
        - ALL

sg_kibana_user:
  cluster:
    - CLUSTER_COMPOSITE_OPS_RO
  indices:
    '?kibana':
      '*':
        - ALL
    '?kibana-devnull':
      '*':
        - KIBANA_READ

sg_own_kibana_index:
  cluster:
    - CLUSTER_COMPOSITE_OPS
  indices:
    '?kibana_${user_name}':
      '*':
        - ALL
    '?kibana-devnull':
      '*':
        - KIBANA_READ
```

sg_roles_mapping.yml:
```
sg_kibana_user:
  users:
    - '*'

sg_own_kibana_index:
  users:
    - '*'
```

kibana.yml:
```
server.port: 5601
server.host: "172.19.6.100"
elasticsearch.url: "http://172.19.6.100:19200"

elasticsearch.username: "kibanaserver"
elasticsearch.password: "kibanaserver"

elasticsearch.requestHeadersWhitelist: [ x-proxy-user, cookie ]
elasticsearch.ssl.verify: false
own_home.elasticsearch.url: "http://172.19.6.100:9200"
own_home.local.groups: [ public ]
```

sg_config.yml
```
searchguard:
  dynamic:
    authc:
      basic_internal_auth_domain: 
        enabled: true
        order: 1
        http_authenticator:
          type: basic
          challenge: true
        authentication_backend:
          type: intern
```

Is it necessary to implement and configure Apache Httpd? Is there any chance to do this without Apache httpd?

**Comment by Poohowy** [comment]
I updated my last post. I added information about my Apache httpd configuration and Kibana error.

**Comment by wtakase** [comment]
In `sg_config.yml`, you need to set 2 authc, `proxy_auth_domain` and `basic_internal_auth_domain` like [this](https://github.com/wtakase/kibana-own-home/issues/12#issuecomment-288646104).

An end-user uses `proxy_auth_domain` because he is authenticated at a http server.
Search Guard receives `x-proxy-user` header and uses it for authorization.
In this case you need to define each username and password in `.htpasswd`.
The web server is an only entry point to access Kibana and Elasticsearch for end-users.

`kibanaserver` uses `basic_internal_auth_domain` because he is not authenticated by the web server.
So you need to define `kibanaserver` and his password in `sg_internal_user.yml`.


In the case that you want to check/test a end-user's permission on Elasticsearch, you can define the username and password in `sg_internal_user.yml`, then access `http://elasticsearchhost:9200` directly with the user's account.

![workflow](https://cloud.githubusercontent.com/assets/8382795/24434021/e1194434-1467-11e7-821d-7dfca4242c9d.png)

**Comment by Poohowy** [comment]
As far as I understand Apache httpd basic auth is something like Search Guard for Kibana. I implemented SG for Kibana and when I open Kibana I see login screen.
![zaloguj](https://cloud.githubusercontent.com/assets/3901058/24447176/624618d6-1470-11e7-8d00-2693e7fbfbb6.PNG)
Tell me if I am wrong, but using httpd and SG Kibana at the same time is problematic, because user have to give username and password twice. 

Is there any way to configure Search Guard for Kibana with Own Home? I wish I could log in using Search Guard for Kibana.

**Comment by wtakase** [comment]
Could you try the following Own Home with Kibana 5.2.2?
[own_home-5.2.2-remore-user.zip](https://github.com/wtakase/kibana-own-home/files/881439/own_home-5.2.2-remore-user.zip)

I don't test enough though, this one supports to extract username from HTTP session instead of `x-proxy-header`.
So you don't need to set up a front end web server.

Install search-guard-kibana-plugin.

Install own home
```
bin/kibana-plugin install https://github.com/wtakase/kibana-own-home/files/881439/own_home-5.2.2-remore-user.zip
```

kibana.yml
```
elasticsearch.url: https://localhost:19200
elasticsearch.requestHeadersWhitelist: [ cookie, authorization ]
elasticsearch.username: kibana
elasticsearch.password: kibana
elasticsearch.ssl.verify: false
server.ssl.cert: /path/to/server.crt
server.ssl.key: /path/to/server.key
console.proxyConfig:
  - match:
      protocol: https
    ssl:
      verify: false 
own_home.get_username_from_session.enabled: true
own_home.get_username_from_session.key: username
own_home.ssl.cert: /path/to/server.crt
own_home.ssl.key: /path/to/server.key
own_home.elasticsearch.url: https://localhost:9200
own_home.session.secretkey: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
searchguard.cookie.secure: true
searchguard.cookie.password: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```

access `https://localhost:5601` and log in as a user defined in `sg_internal_users.yml`.

**Comment by wtakase** [comment]
One more thing.
If you deploy kibana and elasticsearch on different hosts, you may need to apply the following [patch](https://github.com/floragunncom/search-guard-kibana-plugin/pull/3).

**Comment by wtakase** [comment]
Or replace `'_searchguard/authinfo'` with `'/_searchguard/authinfo'` in this [file](https://github.com/floragunncom/search-guard-kibana-plugin/blob/master/lib/backend/searchguard_plugin.js#L29).

**Comment by Poohowy** [comment]
I've got a couple of questions regarding to kibana.yml.

1. What if I did not implement REST layer security, but only transport layer and https is disable? Can I use this parameter:
```
elasticsearch.url: http://localhost:19200
elasticsearch.requestHeadersWhitelist: [ cookie, authorization ]
elasticsearch.username: kibanaserver
elasticsearch.password: kibanaserver
elasticsearch.ssl.verify: false
server.ssl.cert: /path/to/server.crt
server.ssl.key: /path/to/server.key

own_home.get_username_from_session.enabled: true
own_home.get_username_from_session.key: username
own_home.ssl.cert: /path/to/server.crt
own_home.ssl.key: /path/to/server.key
own_home.elasticsearch.url: http://localhost:9200
own_home.session.secretkey: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
searchguard.cookie.secure: true
searchguard.cookie.password: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```
Or do I have to implement REST layer security too?

2. One more question regarding to SSL certifications and keys. Do parameters:
```
server.ssl.cert: /path/to/server.crt
server.ssl.key: /path/to/server.key
own_home.ssl.cert: /path/to/server.crt
own_home.ssl.key: /path/to/server.key
```
have the same value? Or I have to use different certs? 

3. Parameter listed below:
```
server.ssl.cert: /path/to/server.crt
own_home.ssl.cert: /path/to/server.crt
```
are paths to root-ca.cert? Or I should give here path to trustore?

And what about parameters:
```
server.ssl.key: /path/to/server.key
own_home.ssl.key: /path/to/server.key
```
It is node keystore or user keystore?

4. And one more thing. How should look now sgconfig.yml? Is it ok?
```
searchguard:
  dynamic:
    authc:
      basic_internal_auth_domain: 
        enabled: true
        order: 1
        http_authenticator:
          type: basic
          challenge: true
        authentication_backend:
          type: intern
```

Sorry that I have so many questions, but I am newbie to Elastic and SG world. By the way, thanks for your help!

**Comment by wtakase** [comment]
> What if I did not implement REST layer security, but only transport layer and https is disable? Can I use this parameter. Or do I have to implement REST layer security too?

You can configure kibana.yml as follows:
```
elasticsearch.url: http://localhost:19200
elasticsearch.requestHeadersWhitelist: [ cookie, authorization ]
elasticsearch.username: kibanaserver
elasticsearch.password: kibanaserver
elasticsearch.ssl.verify: false
own_home.get_username_from_session.enabled: true
own_home.get_username_from_session.key: username
own_home.session.isSecure: false
own_home.elasticsearch.url: http://localhost:9200
own_home.session.secretkey: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
searchguard.cookie.secure: false
searchguard.cookie.password: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```

> One more question regarding to SSL certifications and keys. Do parameters have the same value? Or I have to use different certs?

You can use the same value or also use different certs.

> Parameter listed below are paths to root-ca.cert? Or I should give here path to trustore?
> And what about parameters, it is node keystore or user keystore?

`ssl.cert` means server's certificate and `ssl.key` means the certificate's key.
https://www.elastic.co/guide/en/kibana/5.2/production.html#enabling-ssl

> And one more thing. How should look now sgconfig.yml? Is it ok?

Yes, it seems ok.

**Comment by Poohowy** [comment]
I changed the configuration and replaced `'_searchguard/authinfo'` with `'/_searchguard/authinfo'` and it worked perfectly!

Thank You very, very much. I really appreciate your help and patience. Thank You!

**Comment by wtakase** [comment]
What about [this v5.3.0-1](https://github.com/wtakase/kibana-own-home/tree/v5.3.0-1)?
And configure as [follows](https://github.com/wtakase/kibana-own-home/tree/v5.3.0-1#extract-username-from-session-instead-of-request-header-experimental)?

**Comment by Poohowy** [comment]
You are right. It works perfect now. Thank you very much.

**Comment by wtakase** [comment]
Sorry, I found a bug that any user could log in...
I fixed it and released new version v5.3.0-2.
